bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -276,6 +283,21 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 }
 
 func (r *KubeadmConfigReconciler) handleClusterNotInitialized(ctx context.Context, scope *Scope) (_ ctrl.Result, reterr error) {
+	// initialize the DataSecretGeneratedCondition if missing.
+	// this is required in order to avoid the condition's LastTransitionTime to flicker in case of errors surfacing
+	// using the DataSecretGeneratedFailedReason
+	if !(conditions.Has(scope.Config, bootstrapv1.DataSecretGeneratedCondition)) {
+		conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretGeneratedCondition, bootstrapv1.WaitingForControlPlaneInitializedReason, clusterv1.ConditionSeverityInfo, "")
+	}
+
+	defer func() {
+		if reterr != nil {
+			// NOTE: We are surfacing errors as a conditions because in some cases those are originated by misconfiguration
+			// that can be fixed by the user. e.g. invalid file source.
+			conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretGeneratedCondition, bootstrapv1.DataSecretGeneratedFailedReason, clusterv1.ConditionSeverityWarning, reterr.Error())
+		}
+	}()
```

Comment:
Would it make sense to push this up to the `Reconcile()` method rather than handling through `defer`?

Eventually, I would expect us to be able to remove all of this special casing (and locking) that we have today around init/join once we can require the use of a control plane provider for managing control plane machines (possibly v1alpha4?)

Reply:
TBH, I'm a little bit debated if to use defer to catch all the errors or if to return an error condition only when there errors due to misconfiguration (e.g. failed to resolve file source).
My main concern is that if we go for "catch-all" approach, we are going to surface a whole set of internal errors that rarely happen, usually recover automatically, and that the user can't really understand or solve.
@vincepri opinions ^^

WRT to have the defer statement in the single reconcile loop or in the main one, I personally prefer to keep them separated, but I don't have strong opinions so I'm happy to change if required

Reply:
I was thinking less around using a defer in `Reconcile()` in favor of handling errors from the individual `reconcile*` methods in `Reconcile()` and setting the conditions accordingly there.